testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/web/security/runas/ServletRunAsTestCase.java

```diff
@@ -0,0 +1,283 @@
+    public void checkDoGetMethodWithIncorrectRole() throws Exception {
+        assertTrue("EJB invocation did not failed in doGet() method of Servlet which uses RunAs with incorrect role.",
+                incorrectRoleResult.contains(DOGET_METHOD_NOT_PASS));
+    }
+
+    /**
+     * Stop server for invoking HttpServlet.destroy() method during undeploying.
+     *
+     * @throws Exception
+     */
+    @Test
+    @InSequence(7)
+    public void runDestroyMethodInUndeploying() throws Exception {
+        deployer.undeploy(DEPLOYMENT);
+    }
+
+    /**
+     * Check whether Servlet which uses RunAs with correct role needed for secured EJB invocation can correctly invoked that EJB
+     * method in HttpServlet.destroy() method during undeploying application.
+     *
+     * @throws Exception
+     */
+    @Test
+    @InSequence(8)
+    public void checkDestroyInUndeployingMethodWithCorrectRole() throws Exception {
+        assertTrue("EJB invocation failed in destroy() method of Servlet which uses RunAs with correct role during undeploying.",
+                readFirstLineOfFile(CORRECT_ROLE_AND_UNDEPLOY).contains(DESTROY_METHOD_PASS + HelloBean.HELLO));
+    }
+
+    /**
+     * Check whether Servlet which uses RunAs with different role than in needed for secured EJB invocation cannot correctly
+     * invoked that EJB method in HttpServlet.destroy() method during undeploying application.
+     *
+     * @throws Exception
+     */
+    @Test
+    @InSequence(9)
+    public void checkDestroyInUndeployingMethodWithIncorrectRole() throws Exception {
+        assertTrue("EJB invocation did not failed in destroy() method of Servlet which uses RunAs with incorrect role during "
+                + "undeploying.",
+                readFirstLineOfFile(INCORRECT_ROLE_AND_UNDEPLOY).contains(DESTROY_METHOD_NOT_PASS));
+    }
+
+    /**
+     * Restart server for invoking HttpServlet.destroy() method during stopping server. It also hit Servlet for initialization
+     * of Servlet before server is restarted.
+     *
+     * @param url
+     * @throws Exception
+     */
+    @Test
+    @InSequence(10)
+    @OperateOnDeployment(DEPLOYMENT)
+    public void runDestroyMethodInStopServer(@ArquillianResource URL url) throws Exception {
+        deployer.deploy(DEPLOYMENT);
+
+        URL servletUrl = new URL(url.toExternalForm() + RunAsAdminServlet.SERVLET_PATH.substring(1) + "?"
+                + Utils.encodeQueryParam(CallProtectedEjbServlet.FILE_PARAM, CORRECT_ROLE_AND_STOP_SERVER.getAbsolutePath()));
+        Utils.makeCall(servletUrl.toURI(), HTTP_OK);
+        servletUrl = new URL(url.toExternalForm() + RunAsUserServlet.SERVLET_PATH.substring(1) + "?"
+                + Utils.encodeQueryParam(CallProtectedEjbServlet.FILE_PARAM, INCORRECT_ROLE_AND_STOP_SERVER.getAbsolutePath()));
+        Utils.makeCall(servletUrl.toURI(), HTTP_OK);
+
+        ServerReload.executeReloadAndWaitForCompletion(managementClient.getControllerClient(), 50000);
```

Comment:
Why?

Is there something special about this scenario that makes it different from all the other undeployment work we do when stopping servers?